src/sgame/sg_bot_parse.cpp

```diff
@@ -1890,3 +1890,295 @@ void FreeBehaviorTree( AIBehaviorTree_t *tree )
+		break;
+	case AIExpType_t::EX_FUNC:
+		{
+			AIValueFunc_t *v = reinterpret_cast<AIValueFunc_t *>( exp );
+			out << SearchFuncName( v );
+			if ( v->nparams > 0 )
+			{
+				out << "( ";
+				for ( int i = 0; i < v->nparams; i++ )
+				{
+					BotValueToString( &v->params[ i ], out );
+					if ( i < v->nparams - 1 )
+					{
+						out << ", ";
+					}
+				}
+				out << " )";
+			}
+		}
+		break;
+	default:
+		out << "<unknown";
+		break;
+	}
+}
+
+static void BotBehaviorToStringRec( AIGenericNode_t *node, std::ostringstream &out, int level )
+{
+	std::string indent = "";
+	for ( int i = 0; i < level; i++ )
+	{
+		indent += "    ";
+	}
+
+	switch ( node->type )
+	{
+	case AINode_t::SPAWN_NODE:
+		{
+			AISpawnNode_t *nd = reinterpret_cast<AISpawnNode_t *>( node );
+			out << indent << "spawnAs " << nd->selection << "\n";
+		}
+		break;
+	case AINode_t::CONDITION_NODE:
+		{
+			AIConditionNode_t *nd = reinterpret_cast<AIConditionNode_t *>( node );
+			out << indent << "condition ";
+			BotExpressionToString( nd->exp, out );
+			if ( nd->child )
+			{
+				out << "\n" << indent << "{\n";
+				BotBehaviorToStringRec( nd->child, out, level + 1 );
+				out << indent << "}";
+			}
+			out << "\n";
+		}
+		break;
+	case AINode_t::DECORATOR_NODE:
+		{
+			AIDecoratorNode_t *nd = reinterpret_cast<AIDecoratorNode_t *>( node );
+			out << indent << "decorator " << SearchDecoratorName( nd );
+			if ( nd->nparams > 0 )
+			{
+				out << "( ";
+				for ( int i = 0; i < nd->nparams; i++ )
+				{
+					BotValueToString( &nd->params[ i ], out );
+					if ( i < nd->nparams - 1 )
+					{
+						out << ", ";
+					}
+				}
+				out << " )";
+			}
+			out << "\n" << indent << "{\n";
+			BotBehaviorToStringRec( nd->child, out, level + 1 );
+			out << indent << "}\n";
+		}
+		break;
+	case AINode_t::BEHAVIOR_NODE:
+		{
+			AIBehaviorTree_t *nd = reinterpret_cast<AIBehaviorTree_t *>( node );
+			out << indent << "behavior " << nd->name << "\n";
+		}
+		break;
+	case AINode_t::ACTION_NODE:
+		{
+			AIActionNode_t *nd = reinterpret_cast<AIActionNode_t *>( node );
+			out << indent << "action " << SearchActionName( nd );
+			if ( nd->nparams > 0 )
+			{
+				out << "( ";
+				for ( int i = 0; i < nd->nparams; i++ )
+				{
+					BotValueToString( &nd->params[ i ], out );
+					if ( i < nd->nparams - 1 )
+					{
+						out << ", ";
+					}
+				}
+				out << " )";
+			}
+			out << "\n";
+		}
+		break;
+	case AINode_t::SELECTOR_NODE:
+		{
+			AINodeList_t *nd = reinterpret_cast<AINodeList_t *>( node );
+			out << indent << SearchSelectorName( nd );
+			out << "\n" << indent << "{\n";
+			for ( int i = 0; i < nd->numNodes; i++ )
+			{
+				BotBehaviorToStringRec( nd->list[ i ], out, level + 1 );
+			}
+			out << indent << "}\n";
+		}
+		break;
+	default:
+		out << indent << "<unknown>\n";
+		break;
+	}
+}
+
+std::string G_BotBehaviorToString( Str::StringRef behavior )
+{
+	AIBehaviorTree_t *tree = BotBehaviorTree( behavior );
```

Comment:
Why don't you use `ReadBehaviorTree` same as when a bot is added?

Reply:
OK so you could write `AIBehaviorTree_t *BotBehaviorTree( Str::StringRef behavior ) { return ReadBehaviorTree( behavior.c_str(), treeList )` in sg_bot.cpp.

Reply:
lol yes